packages/@aws-cdk/aws-lambda/lib/function-base.ts

```diff
@@ -431,6 +435,35 @@ export abstract class FunctionBase extends Resource implements IFunction, ec2.IC
     throw new Error(`Invalid principal type for Lambda permission statement: ${principal.constructor.name}. ` +
       'Supported: AccountPrincipal, ArnPrincipal, ServicePrincipal');
   }
+
+  private parseConditions(principal: iam.IPrincipal): { sourceAccount: string, sourceArn: string } | null {
+    if ('conditions' in principal) {
+      const conditions: iam.Conditions = (principal as iam.PrincipalWithConditions).policyFragment.conditions;
+      const conditionPairs = Object.entries(conditions)
+        .reduce(
+          (accumulator, [operator, conditionObjs]) =>
+            accumulator.concat(Object.keys(conditionObjs as object).map(key => { return { operator, key }; })),
+          ([] as {operator: string, key: string}[]),
+        );
+      const conditionsAreSupported = conditionPairs.every(
+        (condition) => PERMISSION_SUPPORTED_PRINCIPAL_CONDITIONS.some(
+          (supportedCondition) => supportedCondition.operator === condition.operator && supportedCondition.key === condition.key,
+        ),
+      );
+
+      if (conditionsAreSupported) {
+        return {
+          sourceAccount: conditions.StringEquals['aws:SourceAccount'],
+          sourceArn: conditions.ArnLike['aws:SourceArn'],
+        };
+      } else {
+        throw new Error(`PrincipalWithConditions had unsupported conditions for Lambda permission statement: ${JSON.stringify(conditions)}. ` +
```

Comment:
Would be nicer if you can pinpoint exactly what isn't supported.

It's simple enough too, you can filter down the list of conditions:

```ts
const unsupported = conditionPairs.filter(x => !isSupportedCondition(x));

if (unsupported.length > 0) {
  throw new Error(`Oops: ${JSON.stringify(unsupported)}`);
}
```